app/src/main/java/com/panoprogramowanie/testcoroutines/presentation/MainActivity.kt

```diff
@@ -2,30 +2,19 @@ package com.panoprogramowanie.testcoroutines.presentation
 
 import android.app.AlertDialog
 import android.os.Bundle
-import android.support.v7.app.AppCompatActivity
 import com.panoprogramowanie.testcoroutines.DependencyInjector
 import com.panoprogramowanie.testcoroutines.R
 import kotlinx.android.synthetic.main.activity_main.*
-import kotlinx.coroutines.experimental.CoroutineScope
-import kotlinx.coroutines.experimental.Job
 import kotlinx.coroutines.experimental.launch
-import kotlin.coroutines.experimental.CoroutineContext
 
-class MainActivity : AppCompatActivity(), MainView, CoroutineScope {
+class MainActivity : CoroutineScopeActivity(), MainView {
 
-    lateinit var job: Job
-
-    override val coroutineContext: CoroutineContext
-        get() = baseCoroutineScope + job
-
-
-    private val baseCoroutineScope by lazy { DependencyInjector.getUIDispatcher() }
     private val presenter by lazy { DependencyInjector.getMainPresenter() }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
-        presenter.attach(this)
+        launch { presenter.attach(this@MainActivity) }
```

Comment:
@davyskiba the issue I have with this approach is that it breaks MVVM conventions by passing a reference to the Activity to the ViewModel. Is there a way to resolve that?